### PR TITLE
xds: Outlier Detection configuration in Cluster Resolver Balancer

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds branching logic for constructing the Priority LB's config. In the case that Outlier Detection is configured on with the corresponding Environment variable, the top level balancer's configuration for each priority will be of type Outlier Detection with a Cluster Impl child rather than a Cluster Impl. This had to include the Outlier Detection's balancer to pass Cluster Resolver tests, as the Outlier Detection policy needed to be registered.

Branched off #5370.

RELEASE NOTES: None

### CODE REVIEW

File: xds/internal/balancer/clusterresolver/configbuilder.go
Comment (dfawley) [inline xds/internal/balancer/clusterresolver/configbuilder.go:187]: `make(..., len(ciCfgs))`?

File: xds/internal/balancer/clusterresolver/configbuilder.go
Comment (dfawley) [inline xds/internal/balancer/clusterresolver/configbuilder.go:196]: Pass `*clusterimpl.LBConfig` then?

File: xds/internal/balancer/outlierdetection/balancer_test.go
Comment (zasweq) [reply xds/internal/balancer/outlierdetection/balancer_test.go:57]: Good point. This actually lead me to discover that I was actually getting a parsing error not a validation error (see the 100 below, it shouldn't error - and it's also the wrong field name haha).

Comment (zasweq) [reply xds/internal/balancer/outlierdetection/balancer_test.go:57]: Also, another thing I noticed is that I never check if there is a child/the child is a clusterimpl/the child config is valid. These are assumptions I made for my od balancer, which if broken would cause nil panic. What should I do about this? Should I add a validation? Defensively program in my Outlier Detection PR?

File: xds/internal/balancer/outlierdetection/balancer_test.go
Comment (zasweq) [reply xds/internal/balancer/outlierdetection/balancer_test.go:51]: Ah, good point. Switched.

File: xds/internal/balancer/outlierdetection/balancer.go
Comment (zasweq) [reply xds/internal/balancer/outlierdetection/balancer.go:103]: I tried the second option, it broke since this needs to be registered in config validation at a higher level. Thus, I kept implementing balancer.Builder and returned nil from build instead of the type, and deleted this type.

File: xds/internal/balancer/outlierdetection/balancer_test.go
Comment (zasweq) [reply xds/internal/balancer/outlierdetection/balancer_test.go:57]: It won't ever be called otherwise since we prepare the config ourselves but still.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments boss man :D!

File: xds/internal/balancer/clusterresolver/configbuilder.go
Comment (dfawley) [reply xds/internal/balancer/clusterresolver/configbuilder.go]: Pre-allocating is faster so it doesn't need to grow as you add entries if you go over the default initial size.  Or if you have fewer entries than the default initial size, then you save memory.

File: xds/internal/balancer/clusterresolver/configbuilder.go
Comment (dfawley) [reply xds/internal/balancer/clusterresolver/configbuilder.go:191]: Discussed offline

File: xds/internal/balancer/outlierdetection/balancer_test.go
Comment (dfawley) [reply xds/internal/balancer/outlierdetection/balancer_test.go]: Discussed offline

File: xds/internal/balancer/clusterresolver/configbuilder.go
Comment (zasweq) [reply xds/internal/balancer/clusterresolver/configbuilder.go:191]: Fixed it. Let me know if you like solution.

File: xds/internal/balancer/outlierdetection/balancer_test.go
Comment (zasweq) [reply xds/internal/balancer/outlierdetection/balancer_test.go]: Added check for child being there (validating child was already there implicitly from unmarshalJSON if present)

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :D!

File: xds/internal/balancer/clusterresolver/configbuilder.go
Comment (easwars) [reply xds/internal/balancer/clusterresolver/configbuilder.go:189]: See: go/go-style/decisions#pass-values

File: xds/internal/balancer/clusterresolver/configbuilder_test.go
Comment (easwars) [inline xds/internal/balancer/clusterresolver/configbuilder_test.go:1126]: Nit: The `want` suffix does not read as well as a `want` prefix. Maybe s/odCfgsMapWant/wantODCfgs/?

File: xds/internal/balancer/outlierdetection/balancer_test.go
Comment (easwars) [inline xds/internal/balancer/outlierdetection/balancer_test.go:44]: Comment needs fixing.

File: xds/internal/balancer/clusterresolver/configbuilder.go
Comment (zasweq) [reply xds/internal/balancer/clusterresolver/configbuilder.go]: Switched to passing by pointer.

File: xds/internal/balancer/clusterresolver/configbuilder.go
Comment (zasweq) [reply xds/internal/balancer/clusterresolver/configbuilder.go]: Had the discovery mechanism hold onto a value, and pass it by value around. I kept this function returning a pointer though, as that is the value of internalserviceconfig.BalancerConfig that I populate.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments! I'll wait to rebase my full PR since I don't want to force push in case one of you is in the middle of a review on the full one. Maybe I should wait until this is merged to rebase. The balancer is all orthogonal.

File: xds/internal/balancer/clusterresolver/clusterresolver_test.go
Comment (easwars) [inline xds/internal/balancer/clusterresolver/clusterresolver_test.go:148]: Would this work instead?

```golang
if diff := cmp.Diff(gotCCS, wantCCS, cmpopts.IgnoreFields(balancer.ClientConnState{}, "ResolverState")); diff != "" {
...
}
```

Also, why do we have to ignore the resolver state part of the update? Just so that we dont have to specify it in the `wantCCS`?

Comment (zasweq) [reply xds/internal/balancer/clusterresolver/clusterresolver_test.go:148]: Yup, that works :). I just commented this out to test since I forgot since it's been so long, and yeah super verbose resolver state I would have to match in wantCCS.

**Review by zasweq — COMMENTED** [review]
Thanks for the comment :D!

**Review by easwars — APPROVED** [review]

**Pull request merged**
